internal/ansible/metrics/metrics.go

```diff
@@ -82,6 +84,114 @@ func RegisterBuildInfo(r prometheus.Registerer) {
+	Add              float64 `json:"add,omitempty" yaml:"add,omitempty"`
+	Sub              float64 `json:"subtract,omitempty" yaml:"subtract,omitempty"`
+}
+
+type UserMetricHistogram struct {
+	Observe float64 `json:"observe,omitempty" yaml:"observe,omitempty"`
+}
+
+type UserMetricSummary struct {
+	Observe float64 `json:"observe,omitempty" yaml:"observe,omitempty"`
+}
+
+func HandleUserMetric(r prometheus.Registerer, metricSpec UserMetric) error {
+	defer recoverMetricPanic()
+	logf.Log.WithName("metrics").Info("Registering", "metric", metricSpec.Name)
+	_, ok := userMetrics[metricSpec.Name]
+	if !ok {
+		// This is the first time we've seen this metric
+		if metricSpec.Counter != nil {
+			userMetrics[metricSpec.Name] = prometheus.NewCounter(prometheus.CounterOpts{
+				Name: metricSpec.Name,
+				Help: metricSpec.Help,
+			})
+		}
+		if metricSpec.Gauge != nil {
+			userMetrics[metricSpec.Name] = prometheus.NewGauge(prometheus.GaugeOpts{
+				Name: metricSpec.Name,
+				Help: metricSpec.Help,
+			})
+		}
+		if metricSpec.Histogram != nil {
+			userMetrics[metricSpec.Name] = prometheus.NewHistogram(prometheus.HistogramOpts{
+				Name: metricSpec.Name,
+				Help: metricSpec.Help,
+			})
+		}
+		if metricSpec.Summary != nil {
+			userMetrics[metricSpec.Name] = prometheus.NewSummary(prometheus.SummaryOpts{
+				Name: metricSpec.Name,
+				Help: metricSpec.Help,
+			})
+		}
+		r.MustRegister(userMetrics[metricSpec.Name])
+	}
+
+	collector := userMetrics[metricSpec.Name]
+	switch v := collector.(type) {
+	case prometheus.Gauge:
+		if metricSpec.Gauge == nil {
+			return fmt.Errorf("cannot change metric type of %s, which is a gauge", metricSpec.Name)
+		}
+		if metricSpec.Gauge.Inc {
+			v.Inc()
+		} else if metricSpec.Gauge.Dec {
+			v.Dec()
+		} else if metricSpec.Gauge.Add != 0.0 {
+			v.Add(metricSpec.Gauge.Add)
+		} else if metricSpec.Gauge.Sub != 0.0 {
+			v.Sub(metricSpec.Gauge.Sub)
+		} else if metricSpec.Gauge.Set != 0.0 {
+			v.Set(metricSpec.Gauge.Set)
+		} else if metricSpec.Gauge.SetToCurrentTime {
+			v.SetToCurrentTime()
+		}
+	// Counter must be first, because otherwise it can be confused with a gauge.
+	case prometheus.Counter:
+		if metricSpec.Counter == nil {
+			return fmt.Errorf("cannot change metric type of %s, which is a counter", metricSpec.Name)
+		}
+		if metricSpec.Counter.Inc {
+			v.Inc()
+		} else if metricSpec.Counter.Add != 0.0 {
+			v.Add(metricSpec.Counter.Add)
+		}
+	case prometheus.Histogram:
+		if metricSpec.Histogram == nil && metricSpec.Summary == nil {
+			return fmt.Errorf("cannot change metric type of %s, which is a histogram or summary", metricSpec.Name)
+		}
+		// Histogram and Summary interfaces are identical, so we accept either case.
+		if metricSpec.Histogram != nil {
+			v.Observe(metricSpec.Histogram.Observe)
+		}
+		if metricSpec.Summary != nil {
+			v.Observe(metricSpec.Summary.Observe)
+		}
+	}
+	return nil
```

Comment:
Here's my idea, for a refactor of this function that works around the `prometheus.Summary` and `prometheus.Histogram` interface collision. It makes some of my prometheus comments null if you go with it.
```go
func HandleUserMetric(r prometheus.Registerer, metricSpec UserMetric) error {
    if err := validateMetricSpec(metricSpec); err != nil { /* ... */ } // one and only one collector is configured
    collector, err := ensureCollector(metricSpec) // get or create then return collector
    if err != nil { /* ... */ }
    if metricSpec.Counter != nil {
        return handleUserCounter(metricSpec)
    } else if metricSpec.Guage != nil {
        // etc
    }

    return errors.New("...")
}

func handleUserCounter(collector prometheus.Collector, metricSpec UserMetric) error {
    v := collector.(prometheus.Counter) // should probably check the assertion
    if metricSpec.Counter == nil {
 		return fmt.Errorf("cannot change metric type of %s, which is a counter", metricSpec.Name)
 	}
 	if metricSpec.Counter.Inc {
 		v.Inc()
 	} else if metricSpec.Counter.Add != 0.0 {
 		v.Add(metricSpec.Counter.Add)
 	}
    return nil
```

Reply:
Actually, you definitely need to check the assertion. If you POST the same metric name with two different types it'll blow up